wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,370 @@
+
+    createMonitorIfAbsent(props);
+
+    // If needed, wait a short time for custom endpoint info to be discovered.
+    boolean isInfoInCache = waitForCustomEndpointInfo();
+
+    if (!isInfoInCache) {
+      throw new SQLException(
+          Messages.get("CustomEndpointPlugin.cacheTimeout",
+              new Object[]{this.waitOnCachedInfoDurationMs, this.customEndpointHostSpec.getHost()}));
+    }
+
+    return connectFunc.call();
+  }
+
+  /**
+   * Creates a monitor for the custom endpoint if it does not already exist.
+   *
+   * @param props The connection properties.
+   */
+  protected void createMonitorIfAbsent(Properties props) {
+    monitors.computeIfAbsent(
+        this.customEndpointHostSpec.getHost(),
+        (customEndpoint) -> new CustomEndpointMonitorImpl(
+            this.pluginService,
+            this.customEndpointHostSpec,
+            getRegion(this.customEndpointHostSpec, props),
+            TimeUnit.MILLISECONDS.toNanos(CUSTOM_ENDPOINT_INFO_REFRESH_RATE_MS.getLong(props)),
+            this.rdsClientFunc
+        ),
+        TimeUnit.MILLISECONDS.toNanos(this.idleMonitorExpirationMs)
+    );
+  }
+
+  /**
+   * Determines the AWS region from the given parameters. The region specified by the props will be used if it is
+   * provided. Otherwise, the region will be parsed from the specified host URL.
+   *
+   * @param hostSpec Host information for the connection being established.
+   * @param props    The connection properties for the connection being established.
+   * @return The AWS region of the connection being established.
+   */
+  protected Region getRegion(HostSpec hostSpec, Properties props) {
+    String regionString = REGION_PROPERTY.getString(props);
+    if (StringUtils.isNullOrEmpty(regionString)) {
+      regionString = rdsUtils.getRdsRegion(hostSpec.getHost());
+    }
+
+    if (StringUtils.isNullOrEmpty(regionString)) {
+      throw new
+          RuntimeException(
+          Messages.get(
+              "CustomEndpointMonitorImpl.missingRequiredConfigParameter",
+              new Object[] {REGION_PROPERTY.name}));
+    }
+
+    final Region region = Region.of(regionString);
+    if (!Region.regions().contains(region)) {
+      throw new RuntimeException(
+          Messages.get(
+              "AwsSdk.unsupportedRegion",
+              new Object[] {regionString}));
+    }
+
+    return region;
+  }
+
+  /**
+   * If custom endpoint info does not exist for the current custom endpoint, waits a short time for the info to be
+   * made available by the custom endpoint monitor. This is necessary so that other plugins can rely on accurate custom
+   * endpoint info. Since custom endpoint monitors and information are shared, we should not have to wait often.
+   *
+   * @return true if the custom endpoint info is available, or false if we timed out while waiting for the info to
+   *     become available.
+   */
+  protected boolean waitForCustomEndpointInfo() {
```

Comment:
What if a user doesn't want to wait? Can we not wait and proceed with a custom endpoint and let DNS resolve it for us?

Reply:
The problem with not waiting is that, if the custom endpoint info isn't discovered yet, there will be no disallowed hosts. Which means that failover and the read write splitting plugin may connect to hosts outside of the custom endpoint. It seems like the user would not want this to happen.